Unable to call composite recursive function

The following program `sample.go` triggers a panic:

```go
package main

import "fmt"

type A struct {
	B string
	C D
}

func (a *A) Test() string {
	return "test"
}

type D struct {
	E *A
}

func main() {
	a := &A{B: "b"}
	d := D{E: a}
	fmt.Println(d.E.Test())
}
```

Expected result:
```console
$ go run ./sample.go
test
```

Got:
```console
$ yaegi ./sample.go
_test/struct47.go:19:10: panic
reflect.Set: value of type *interface {} is not assignable to type *struct { B string; C struct { E *interface {} } }
goroutine 1 [running]:
runtime/debug.Stack(0x1, 0xc000326000, 0x40)
	/usr/local/Cellar/go/1.14.3/libexec/src/runtime/debug/stack.go:24 +0x9d
github.com/containous/yaegi/interp.(*Interpreter).Eval.func1(0xc000155df0)
	/Users/nrwiersma/Go/src/github.com/containous/yaegi/interp/interp.go:328 +0xbc
panic(0x1897d20, 0xc000186b20)
	/usr/local/Cellar/go/1.14.3/libexec/src/runtime/panic.go:969 +0x166
github.com/containous/yaegi/interp.runCfg.func1(0xc0002f20a0, 0xc00030b300)
	/Users/nrwiersma/Go/src/github.com/containous/yaegi/interp/run.go:113 +0x20c
panic(0x1897d20, 0xc000186b20)
	/usr/local/Cellar/go/1.14.3/libexec/src/runtime/panic.go:969 +0x166
reflect.Value.assignTo(0x187b680, 0xc0001902a8, 0x196, 0x19cad34, 0xb, 0xc00018c780, 0x0, 0xc00018c980, 0x101, 0xc0001902e8)
	/usr/local/Cellar/go/1.14.3/libexec/src/reflect/value.go:2403 +0x426
reflect.Value.Set(0xc00018c780, 0xc0001902e8, 0x196, 0x187b680, 0xc0001902a8, 0x196)
	/usr/local/Cellar/go/1.14.3/libexec/src/reflect/value.go:1532 +0xbd
github.com/containous/yaegi/interp.call.func7(0xc0002f20a0, 0xc0002f18f0)
	/Users/nrwiersma/Go/src/github.com/containous/yaegi/interp/run.go:821 +0x843
github.com/containous/yaegi/interp.runCfg(0xc00030b300, 0xc0002f20a0)
	/Users/nrwiersma/Go/src/github.com/containous/yaegi/interp/run.go:119 +0x6d
github.com/containous/yaegi/interp.(*Interpreter).run(0xc0002ee000, 0xc00030a800, 0xc0002f2000)
	/Users/nrwiersma/Go/src/github.com/containous/yaegi/interp/run.go:97 +0x239
github.com/containous/yaegi/interp.(*Interpreter).Eval(0xc0002ee000, 0xc0002fc000, 0xe1, 0x0, 0x0, 0x0, 0x0, 0x0)
	/Users/nrwiersma/Go/src/github.com/containous/yaegi/interp/interp.go:404 +0x52b
main.main()
	/Users/nrwiersma/Go/src/github.com/containous/yaegi/cmd/yaegi/yaegi.go:143 +0x553
```

--- a/_test/struct47.go
+++ b/_test/struct47.go
@@ -1,0 +1,26 @@
+package main
+
+import "fmt"
+
+type A struct {
+	B string
+	C D
+}
+
+func (a *A) Test() string {
+	return "test"
+}
+
+type D struct {
+	E *A
+}
+
+func main() {
+	a := &A{B: "b"}
+	d := D{E: a}
+	a.C = d
+	fmt.Println(a.C.E.Test())
+}
+
+// Output:
+// test

--- a/_test/struct48.go
+++ b/_test/struct48.go
@@ -1,0 +1,35 @@
+package main
+
+type List struct {
+	Next *List
+	Num  int
+}
+
+func add(l *List, n int) *List {
+	if l == nil {
+		return &List{Num: n}
+	}
+	l.Next = add(l.Next, n)
+	return l
+}
+
+func pr(l *List) {
+	if l == nil {
+		println("")
+		return
+	}
+	print(l.Num, " ")
+	pr(l.Next)
+}
+
+func main() {
+	a := add(nil, 0)
+	pr(a)               // so far so good
+	a = add(a, 1)       // fails here
+	pr(a)
+	a = add(a, 2)
+	pr(a)
+}
+
+// Output:
+// 0

--- a/_test/struct48.go
+++ b/_test/struct48.go
@@ -18,7 +18,7 @@ func pr(l *List) {
 		println("")
 		return
 	}
-	print(l.Num, " ")
+	print(l.Num)
 	pr(l.Next)
 }
 
@@ -33,5 +33,5 @@ func main() {
 
 // Output:
 // 0
-// 0  1
-// 0  1  2
+// 01
+// 012

--- a/_test/struct48.go
+++ b/_test/struct48.go
@@ -24,8 +24,8 @@ func pr(l *List) {
 
 func main() {
 	a := add(nil, 0)
-	pr(a)               // so far so good
-	a = add(a, 1)       // fails here
+	pr(a)
+	a = add(a, 1)
 	pr(a)
 	a = add(a, 2)
 	pr(a)
@@ -33,3 +33,5 @@ func main() {
 
 // Output:
 // 0
+// 0  1
+// 0  1  2

--- a/interp/cfg.go
+++ b/interp/cfg.go
@@ -495,7 +495,7 @@ func (interp *Interpreter) cfg(root *node, pkgID string) ([]*node, error) {
 				// Propagate type
 				// TODO: Check that existing destination type matches source type
 				switch {
-				case n.action == aAssign && src.action == aCall && dest.typ.cat != interfaceT:
+				case n.action == aAssign && src.action == aCall && dest.typ.cat != interfaceT && !isRecursiveField(dest):
 					// Call action may perform the assignment directly.
 					n.gen = nop
 					src.level = level
@@ -1994,6 +1994,10 @@ func isKey(n *node) bool {
 
 func isField(n *node) bool {
 	return n.kind == selectorExpr && len(n.child) > 0 && n.child[0].typ != nil && isStruct(n.child[0].typ)
+}
+
+func isRecursiveField(n *node) bool {
+	return isField(n) && (n.typ.recursive || n.typ.cat == ptrT && n.typ.val.recursive)
 }
 
 // isNewDefine returns true if node refers to a new definition

--- a/interp/run.go
+++ b/interp/run.go
@@ -662,6 +662,8 @@ func call(n *node) {
 			}
 			if len(n.child[0].typ.arg) > i && n.child[0].typ.arg[i].cat == interfaceT {
 				values = append(values, genValueInterface(c))
+			} else if isRecursiveStruct(c.typ, c.typ.rtype) {
+				values = append(values, genValueDerefInterfacePtr(c))
 			} else {
 				values = append(values, genValue(c))
 			}
@@ -827,7 +829,10 @@ func call(n *node) {
 						vararg.Set(reflect.Append(vararg, v(f)))
 					}
 				default:
-					dest[i].Set(v(f))
+					val := v(f)
+					if !val.IsZero() {
+						dest[i].Set(val)
+					}
 				}
 			}
 		}

--- a/interp/run.go
+++ b/interp/run.go
@@ -660,11 +660,12 @@ func call(n *node) {
 				}
 				convertLiteralValue(c, argType)
 			}
-			if len(n.child[0].typ.arg) > i && n.child[0].typ.arg[i].cat == interfaceT {
+			switch {
+			case len(n.child[0].typ.arg) > i && n.child[0].typ.arg[i].cat == interfaceT:
 				values = append(values, genValueInterface(c))
-			} else if isRecursiveStruct(c.typ, c.typ.rtype) {
+			case isRecursiveStruct(c.typ, c.typ.rtype):
 				values = append(values, genValueDerefInterfacePtr(c))
-			} else {
+			default:
 				values = append(values, genValue(c))
 			}
 		}

--- a/interp/run.go
+++ b/interp/run.go
@@ -617,7 +617,11 @@ func call(n *node) {
 	var values []func(*frame) reflect.Value
 	if n.child[0].recv != nil {
 		// Compute method receiver value.
-		values = append(values, genValueRecv(n.child[0]))
+		if isRecursiveStruct(n.child[0].recv.node.typ, n.child[0].recv.node.typ.rtype) {
+			values = append(values, genValueRecvInterfacePtr(n.child[0]))
+		} else {
+			values = append(values, genValueRecv(n.child[0]))
+		}
 		method = true
 	} else if n.child[0].action == aMethod {
 		// Add a place holder for interface method receiver.

--- a/interp/value.go
+++ b/interp/value.go
@@ -54,6 +54,25 @@ func genValueRecv(n *node) func(*frame) reflect.Value {
 	}
 }
 
+func genValueRecvInterfacePtr(n *node) func(*frame) reflect.Value {
+	v := genValue(n.recv.node)
+	fi := n.recv.index
+
+	return func(f *frame) reflect.Value {
+		r := v(f)
+		r = r.Elem().Elem()
+
+		if len(fi) == 0 {
+			return r
+		}
+
+		if r.Kind() == reflect.Ptr {
+			r = r.Elem()
+		}
+		return r.FieldByIndex(fi)
+	}
+}
+
 func genValueAsFunctionWrapper(n *node) func(*frame) reflect.Value {
 	value := genValue(n)
 	typ := n.typ.TypeOf()

--- a/interp/value.go
+++ b/interp/value.go
@@ -195,6 +195,18 @@ func genValueInterface(n *node) func(*frame) reflect.Value {
 	}
 }
 
+func genValueDerefInterfacePtr(n *node) func(*frame) reflect.Value {
+	value := genValue(n)
+
+	return func(f *frame) reflect.Value {
+		v := value(f)
+		if v.IsZero() {
+			return v
+		}
+		return v.Elem().Elem()
+	}
+}
+
 func zeroInterfaceValue() reflect.Value {
 	n := &node{kind: basicLit, typ: &itype{cat: nilT, untyped: true}}
 	v := reflect.New(interf).Elem()

